4091 - Add truncated style to Breadcrumbs

EdwardCoyle opened the pull request:
**Explain the _details_ for making this change. What existing problem does the pull request solve?**
This PR:
- Adds a truncated style to the Breadcrumb component, per request from @gushanson
- Sets this new style to the default (old wrapping style is still optionally available / used as a fallback in IE11)
- Makes various visual tweaks to all Breadcrumb variants (changes in padding/spacing/etc)
- Adds a missing API endpoint (`get disabled() on individual breadcrumbs`)
- Adds/updates tests for the new truncation feature

**Related github/jira issue (required)**:
Closes #4091

**Steps necessary to review your pull request (required)**:
Pull this branch, build, run the demoapp. Then test all scenarios below.  On all scenarios, also all combinations of RTL, Uplift, Personalization (all should be working).

**NOTE:** The resize detection was created using a `ResizeObserver`.  This primitive doesn't exist in IE11, so if you test this feature in IE11, the breadcrumb list does not truncate.  However, it should gracefully degrade back to the original, non-truncated style in this case, so it's worth testing.

_Test basic samples:_
- Open http://localhost:4000/components/breadcrumb/example-index.html
- When this page loads at a large form-factor, no truncation should occur.
- Also open http://localhost:4000/components/breadcrumb/example-from-settings.html
- When this page loads at a large form-factor, no truncation should occur.
- Clicking Destroy/Invoke should work without causing any JS console errors.

_Test truncation:_
- Open http://localhost:4000/components/breadcrumb/example-add-remove.html
- Click the "Add" button to add enough breadcrumbs to the list to cause truncation.
- Click the "..." button that appears to the left of the breadcrumbs to open the overflow menu.  When it opens, it should display all list items that are cut off and/or hidden from the list.
- Click any of the items in the overflow menu.  When clicking, a toast should display with the correct Breadcrumb name (this signals that the Breadcrumb item's callback method is properly being triggered from the overflow menu).
- Try adding/removing breadcrumbs, adding disabled breadcrumbs, and using the `Tab`/`Shift`+`Tab` keys to navigate across the Breadcrumb list. Any items that are cut off/hidden/disabled should not become focused.  The Popupmenu should open using its normal keyboard handling.

_Test header styles:_
Try the following pages to test the standard Header breadcrumb style:
- http://localhost:4000/components/breadcrumb/example-navigation-breadcrumbs.html
- http://localhost:4000/components/breadcrumb/example-navigation-breadcrumbs.html?theme=uplift
http://localhost:4000/components/breadcrumb/example-navigation-breadcrumbs.html?theme=uplift&variant=dark&colors=7928E1

Try the following pages to test the alternate Header breadcrumb style:
- http://localhost:4000/components/breadcrumb/example-navigation-breadcrumbs-alternate.html
- http://localhost:4000/components/breadcrumb/example-navigation-breadcrumbs-alternate.html?theme=uplift
http://localhost:4000/components/breadcrumb/example-navigation-breadcrumbs-alternate.html?theme=uplift&variant=dark&colors=7928E1

Try various combinations of themes/personalization.

_Test where `truncate: false` is set:_
- Open http://localhost:4000/components/breadcrumb/example-no-truncate.html
- Observe this breadcrumb list will always wrap and never truncate.

**Included in this Pull Request**:
- [x] An e2e or functional test for the bug or feature.
- [x] A note to the change log.

tmcconechy commented on docs/CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 
 ### v4.32.0 Features
 
+- `[Breadcrumb]` Add truncated style and made it the default for all Breadcrumb lists. `EPC` ([#4091](https://github.com/infor-design/enterprise/issues/4091))
```
```suggestion
- `[Breadcrumb]` Add truncated style and made it the default for all Breadcrumb lists. ([#4091](https://github.com/infor-design/enterprise/issues/4091))
```

tmcconechy commented on docs/CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 
+- `[Breadcrumb]` Add truncated style and made it the default for all Breadcrumb lists. `EPC` ([#4091](https://github.com/infor-design/enterprise/issues/4091))
 - `[Datagrid]` Add a new `RowNumber` formatter that will show a row number column that remains the same no matter how the grid is sorted. ([#1904](https://github.com/infor-design/enterprise/issues/1904))
 - `[Datepicker]` Added the ability to use the range selection in date picker when using the UmAlQura Calendar (RTL). `TJM` ([#4227](https://github.com/infor-design/enterprise/issues/4227))
```
```suggestion
- `[Datepicker]` Added the ability to use the range selection in date picker when using the UmAlQura Calendar (RTL). ([#4227](https://github.com/infor-design/enterprise/issues/4227))
```

tmcconechy commented on src/components/breadcrumb/breadcrumb.js:
```diff
@@ -104,6 +107,23 @@ BreadcrumbItem.prototype = {
+   * Set the anchor's tabIndex based on its current overflow state, if applicable
+   * @returns {void}
+   */
+  checkFocus() {
```
Is this private?

EdwardCoyle replied:
This one probably should be

tmcconechy commented on src/components/breadcrumb/breadcrumb.js:
```diff
@@ -104,6 +107,23 @@ BreadcrumbItem.prototype = {
+  /**
+   * @returns {HTMLAnchorElement} reference to this breadcrumb item's anchor tag
+   */
+  get a() {
```
And this?

EdwardCoyle replied:
I actually used this one in the tests, I think it's OK to expose if you've got access to the breadcrumb API

pwpatton commented on src/components/breadcrumb/breadcrumb.js:
```diff
@@ -512,6 +625,31 @@ Breadcrumb.prototype = {
+        const liItem = args[0];
+        const index = liItem[0].getAttribute('data-breadcrumb-index');
+        const breadcrumbAPI = this.overflowed[Number(index)];
+        $(breadcrumbAPI.a).trigger('click');
```
Nevermind, I debugged and found it all goes through the callback :)

pwpatton commented:
This looks great!

deep7102 commented:
This looks so good! found few also as:

need to fix this destroy example page, because destroy not working here http://localhost:4000/components/breadcrumb/example-destroy-invoke.html but working here http://localhost:4000/components/breadcrumb/example-from-settings.html

safari - cannot get to breadcrumbs by keyboard
http://localhost:4000/components/breadcrumb/example-callback.html

ie11 - can not load breadcrumbs
http://localhost:4000/components/breadcrumb/example-callback.html

ie11 and edge - can not get to first breadcrumb (Home) by keyboard
http://localhost:4000/components/breadcrumb/example-index.html

EdwardCoyle commented:
@deep7102 I think I addressed the issues with tabIndex.  As far as Safari goes, make sure the setting for respecting all tabIndexes is active in your Safari Preferences.  I wasn't seeing that issue until I tried turning it off:

<img width="886" alt="Screen Shot 2020-08-27 at 11 53 50 AM" src="https://user-images.githubusercontent.com/3249601/91465710-21cdeb80-e85c-11ea-8b69-24418b49f4bb.png">

Regarding the "remaining" breadcrumbs on the page here: http://localhost:4000/components/breadcrumb/example-destroy-invoke.html -- this is actually a built in feature for backwards compatibility.  If you generate Breadcrumbs via HTML markup instead of by doing it with a dataset, the ones generated from HTML hang around until explicitly removed.  On the page where all the breadcrumbs disappear, all of them are generated from data instead.
